pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,295 @@
+	// 2. for each TopologySpreadyConstraint in that namespace
+	//  { find all evictable pods in that namespace
+	//  { 3. for each evictable pod in that namespace
+	// 4. If the pod matches this TopologySpreadConstraint LabelSelector
+	// 5. If the pod nodeName is present in the nodeMap
+	// 6. create a topoPair with key as this TopologySpreadConstraint.TopologyKey and value as this pod's Node Label Value for this TopologyKey
+	// 7. add the pod with key as this topoPair
+	// 8. find the min number of pods in any topoPair for this topologyKey
+	// iterate through all topoPairs for this topologyKey and diff currentPods -minPods <=maxSkew
+	// if diff > maxSkew, add this pod in the current bucket for eviction
+
+	// First record all of the constraints by namespace
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Couldn't list namespaces")
+		return
+	}
+	podsForEviction := make(map[*v1.Pod]struct{})
+	// 1. for each namespace...
+	for _, namespace := range namespaces.Items {
+		if (!includedNamespaces.Has(namespace.Name) || excludedNamespaces.Has(namespace.Name)) && (includedNamespaces.Len()+excludedNamespaces.Len() > 0) {
+			continue
+		}
+		namespacePods, err := client.CoreV1().Pods(namespace.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			klog.ErrorS(err, "Couldn't list pods in namespace", "namespace", namespace)
+			continue
+		}
+
+		// ...where there is a topology constraint
+		//namespaceTopologySpreadConstrainPods := make([]v1.Pod, 0, len(namespacePods.Items))
+		namespaceTopologySpreadConstraints := make(map[v1.TopologySpreadConstraint]struct{})
+		for _, pod := range namespacePods.Items {
+			for _, constraint := range pod.Spec.TopologySpreadConstraints {
+				namespaceTopologySpreadConstraints[constraint] = struct{}{}
+			}
+		}
+		if len(namespaceTopologySpreadConstraints) == 0 {
+			return
+		}
+
+		// 2. for each topologySpreadConstraint in that namespace
+		for constraint := range namespaceTopologySpreadConstraints {
+			// Only deal with hard topology constraints
+			// TODO(@damemi): add support for soft constraints
+			if constraint.WhenUnsatisfiable != v1.DoNotSchedule {
+				continue
+			}
+			constraintTopologies := make(map[topologyPair][]*v1.Pod)
+			// pre-populate the topologyPair map with all the topologies available from the nodeMap
+			// (we can't just build it from existing pods' nodes because a topology may have 0 pods)
+			for _, node := range nodeMap {
+				if val, ok := node.Labels[constraint.TopologyKey]; ok {
+					constraintTopologies[topologyPair{key: constraint.TopologyKey, value: val}] = make([]*v1.Pod, 0)
+				}
+			}
+
+			// 3. for each evictable pod in that namespace
+			// (this loop is where we count the number of pods per topologyValue that match this constraint's selector)
+			var sumPods float64
+			for i := range namespacePods.Items {
+				if !evictable.IsEvictable(&namespacePods.Items[i]) {
+					continue
```

Comment:
Unevictable pods are ignored while their constraints not, then pod-num in some topology domains may be less than it actually is. Suppose we have two domains A(1 pod) and B(3 Pods), maxSkew is 1, 1 pod on domain B should be evicted, but if all these pods except one pod on B are unevictable, no pod will be evicted because maxSkew hasn't be reached.
I think we can treat unevictable pods like pods with nodeAffinity, only ignore them when we try to evict them.

Reply:
I don't think the descheduler should ever consider unevictable pods as evictable, even in this case. Though it is worth noting in documentation that unevictable pods may hurt the effectiveness of this strategy, that's about the best we can do.

imo they are unevictable for a reason and we provide various knobs for the user to adjust that criteria

Reply:
@lixiang233 revisiting this, after running the strategy in more live-cluster scenarios I apologize and take back what I said. I agree with you that we shouldn't be excluding unevictable pods at this point, but rather when we try to evict them. That's much more intuitive and you were absolutely correct!

Reply:
However we should sort non-evictable pods to be considered last for eviction, so I think it could be helpful to make the [`evictable`](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/evictions/evictions.go#L170) type exported so that I can pass it to my sort functions here. @ingvagabund wdyt?